examples/create_mask.rs

```diff
@@ -75,104 +89,141 @@ fn main() {
-				imgproc::fill_poly(
-					&mut mask,
-					&points_mat,
-					Scalar::new(255., 255., 255., 255.),
+				let point = Point::new(x, y);
+				imgproc::circle(
+					&mut next_frame,
+					point,
+					2,
+					Scalar::new(0., 0., 255., 0.),
+					-1,
 					imgproc::LINE_8,
 					0,
-					Point::default(),
 				)
 				.unwrap();
+				marker_points.push(point);
+
+				if marker_points.len() > 1 {
+					imgproc::line(
+						&mut next_frame,
+						marker_points[marker_points.len() - 2],
+						point,
+						Scalar::new(0., 0., 255., 0.),
+						2,
+						imgproc::LINE_8,
+						0,
+					)
+					.unwrap();
+				}
 
-				bitwise_and(&src, &src, &mut final_img, &mask).unwrap();
-
-				imshow("Mask", &mask).unwrap();
-				imshow("Result", &final_img).unwrap();
-				imshow(SOURCE_WINDOW, &img1).unwrap();
+				imshow(SOURCE_WINDOW, &next_frame).unwrap();
+			}
+			DrawingState::DrawingMask => {
+				if !marker_points.is_empty() {
+					next_frame = src.clone();
+
+					let pts_mat = Mat::from_slice(marker_points.as_slice()).unwrap();
+					imgproc::polylines(
+						&mut next_frame,
+						&pts_mat,
+						true,
+						Scalar::new(0., 0., 0., 0.),
+						2,
+						imgproc::LINE_8,
+						0,
+					)
+					.unwrap();
+
+					imshow(SOURCE_WINDOW, &next_frame).unwrap();
+				}
+			}
+			DrawingState::DrawingMaskFinished => {
+				if !marker_points.is_empty() {
+					final_img = Mat::zeros_size(src.size().unwrap(), CV_8UC3).unwrap().to_mat().unwrap();
+					mask = Mat::zeros_size(src.size().unwrap(), CV_8UC1).unwrap().to_mat().unwrap();
+
+					imgproc::fill_poly(
+						&mut mask,
+						&Mat::from_slice(marker_points.as_slice()).unwrap(),
+						Scalar::new(255., 255., 255., 255.),
+						imgproc::LINE_8,
+						0,
+						Point::default(),
+					)
+					.unwrap();
+
+					bitwise_and(&src, &src, &mut final_img, &mask).unwrap();
+
+					imshow("Mask", &mask).unwrap();
+					imshow("Result", &final_img).unwrap();
+					imshow(SOURCE_WINDOW, &next_frame).unwrap();
+				}
 			}
-			EVENT_MBUTTONDOWN => {
-				dbg!((mouse_event, x, y));
+			DrawingState::Resetting => {
+				if !marker_points.is_empty() {
+					marker_points.clear();
+					next_frame = src.clone();
 
-				num_points_has_add = 0;
-				points.clear();
-				img1 = src.clone();
-				imshow(SOURCE_WINDOW, &img1).unwrap();
+					imshow(SOURCE_WINDOW, &next_frame).unwrap();
+				}
 			}
-			_ => {}
+		}
+	}
+}
+
+/// Converts an `i32` to a `opencv::highgui::MouseEventTypes`
+///
+/// # Panics
+///
+/// Panics if the argument less than 0 or greater than 11.
+fn mouse_event_from_i32(value: i32) -> opencv::highgui::MouseEventTypes {
+	(value.gt(&(opencv::highgui::MouseEventTypes::EVENT_MOUSEHWHEEL as i32) /* 11 */)
+		|| (value.lt(&(opencv::highgui::MouseEventTypes::EVENT_MOUSEMOVE as i32) /* 0 */)))
+	.then(|| panic!("Invalid cv::highgui::MouseEventTypes value: {}", value));
+
+	// Safe because of the previous check
+	unsafe { std::mem::transmute(value) }
+}
```

Comment:
Can this function be generated automatically during the build process?

Reply:
Definitely! I’ll take a look.

Reply:
Thank you!
- With this automatic implementation, we can easily use `highgui::MouseEventTypes` enum and pattern matching together to take advantage of exhaustiveness checking, enabling us to discover issues earlier. This approach is preferred over comparing values of type i32 with if-else statements and certain constants(e.g. `if event == highgui::EVENT_MOUSEMOVE {} `).

  I think the same  principle can be applied to other enums within  `opencv::highgui` module, such as  `highgui::MouseEventFlags` 、`highgui::QtButtonTypes ` etc.

  ```rust
  impl From<i32> for crate::highgui::MouseEventTypes {
  	fn from(v: i32) -> Self {
  		(v.gt(&(crate::highgui::MouseEventTypes::EVENT_MOUSEHWHEEL as i32))
  			|| (v.lt(&(crate::highgui::MouseEventTypes::EVENT_MOUSEMOVE as i32))))
  		.then(|| panic!("Invalid cv::highgui::MouseEventTypes value: {}", v));
  		// Safe because of the previous check
  		unsafe { std::mem::transmute(v) }
  	}
  }
  ```

- And the type `MouseCallback` may can be redefined

  ```rust
  pub type MouseCallback = Option<Box<dyn FnMut(i32, i32, i32, i32) + Send + Sync + 'static>>;
  ```

  to

  ```rust
  pub type MouseCallback = Option<Box<dyn FnMut(highgui::MouseEventTypes, i32, i32, highgui::MouseEventFlags) + Send + Sync + 'static>>;
  ```

  But it will make breaking changes!